deepspeed/pt/deepspeed_zero_optimizer.py

```diff
@@ -182,10 +222,26 @@ def __init__(self,
         #each of which will be updated by a different process
         self.parallel_partitioned_fp16_groups = []
 
+        #jie:
         #a single 32-bit partition of the parallel partitioned parameters
         #that this process will update
-        self.single_partition_of_fp32_groups = []
 
+        self.total_params = 0
+        self.single_partition_of_fp32_groups_index = 0
+
+        if self.cpu_offload:
+            logger.info(f"cpu_offload enabled")
+            for i, param_group in enumerate(self.optimizer.param_groups):
+                self.total_params += sum([t.numel() for t in param_group['params']])
+            logger.info("Total # of params is {}".format(self.total_params))
+            self.single_partition_of_fp32_groups = torch.zeros(

```

Comment:
I'm not close to the ZeRO codebase at all, so please correct me whenever applicable :-).

To my understanding, a lot of the ways that the CPU branches of the code diverge from the regular code path comes from `self.single_partition_of_fp32_groups` being a flattened buffer instead of a list of tensors.

I think it may be tricky to maintain/test/debug some of this design because the variable can be a different data structure depending on our configuration.

I get the performance advantages of keeping things in a nice flattened buffer. Would it be possible to instead allocate the flattened buffer and then make `self.single_partition_of_fp32_groups` a list of tensors that view the single buffer? Then there would be fewer code changes because it's the same data structure end-to-end and we're just changing the backing storage. I find this stuff super tricky and there may be some computation graph or other major consequences that I am missing.